### PR TITLE
Slackv3 HotFix for SocketMode concurrency bug.

### PR DESCRIPTION
**Pull request by amshamah419**
## Status
- [ ] In Progress
- [x] Ready
- [ ] In Hold - (Reason for hold)

## Related Issues
fixes: https://github.com/demisto/etc/issues/39986

## Description
Long-running was not utilizing a lock for the demisto calls. This PR implements the CommonServer function to handle the lock acquisition and releasing of the demisto object. There are some changes to the threading to enable the locking to work.

Additionally, some scopes were missing as identified by some customers. Those have been verified and added to the App Manifest.

### CODE REVIEW

File: Packs/Slack/Integrations/SlackV3/SlackV3.py [hunk 1]
```diff
@@ -588,24 +588,21 @@ def mirror_investigation():
...
 
-def long_running_loop():
+async def long_running_loop():
     while True:
         error = ''
         try:
-            check_for_mirrors()
-            check_for_unanswered_questions()
+            integration_context = check_for_mirrors()
+            check_for_unanswered_questions(integration_context)
```
Comment (amshamah419) [inline Packs/Slack/Integrations/SlackV3/SlackV3.py:596]: Reducing calls to `demisto.getIntegrationContext()` and passing them between the two. Every time a `demisto` object is called, we lock and unlock the resource so to reduce the need, I'm limiting the number of calls.

File: Packs/Slack/Integrations/SlackV3/SlackV3.py [hunk 2]
```diff
@@ -1890,7 +1907,14 @@ def long_running_main():
...
     """
-    asyncio.run(start_listening())
+    try:
+        asyncio.run(start_listening(), debug=True)
+    except Exception as e:
+        demisto.error(f"The Loop has failed to run {str(e)}")
+    finally:
+        loop = asyncio.get_running_loop()
+        loop.stop()
+        loop.close()
```
Comment (amshamah419) [inline Packs/Slack/Integrations/SlackV3/SlackV3.py:1917]: Wrap in try-catch to catch loop exception

File: Packs/Slack/ReleaseNotes/2_1_3.md [hunk 3]
```diff
@@ -0,0 +1,5 @@
+
+#### Integrations
+##### Slack v3 (beta)
+- Fixed an issue where SocketMode would fail to connect.
+- Improved the App Manifest to contain more OAuth scopes for functionality.
```
Comment (yuvalbenshalom) [inline Packs/Slack/ReleaseNotes/2_1_3.md:5]: probably needs an update

File: Packs/Slack/Integrations/SlackV3/SlackV3.py [hunk 4]
```diff
@@ -989,10 +1010,13 @@ async def create_incidents(incidents: list, user_name: str, user_email: str, use
...
 async def listen(client: SocketModeClient, req: SocketModeRequest):
+    if req:
+        demisto.info("Handling request")
     data_type: str = req.type
     payload: dict = req.payload
-    response = SocketModeResponse(envelope_id=req.envelope_id)
-    await client.send_socket_mode_response(response)
+    if req.envelope_id:
+        response = SocketModeResponse(envelope_id=req.envelope_id)
+        await client.send_socket_mode_response(response)
```
Comment (yuvalbenshalom) [inline Packs/Slack/Integrations/SlackV3/SlackV3.py:1019]: I think line 1015-1019 should all be under the scope of `if req:`

File: Packs/Slack/Integrations/SlackV3/SlackV3.py [hunk 5]
```diff
@@ -787,30 +786,53 @@ def extract_entitlement(entitlement: str, text: str) -> Tuple[str, str, str, str
...
-        app_token=APP_TOKEN,
-        web_client=ASYNC_CLIENT,
-        logger=slack_logger
-    )
-    client.socket_mode_request_listeners.append(listen)
-    await client.connect()
-    await asyncio.sleep(float("inf"))
+    while True:
+        # SocketModeClient does not respect environment variables for ssl verification.
+        # Instead we use a custom session.
```
Comment (yuvalbenshalom) [inline Packs/Slack/Integrations/SlackV3/SlackV3.py:821]: let's talk about this.

File: Packs/Slack/Integrations/SlackV3/SlackV3.py [hunk 6]
```diff
@@ -787,30 +786,53 @@ def extract_entitlement(entitlement: str, text: str) -> Tuple[str, str, str, str
 
 
 class SlackLogger:
+    """
+    Slack's Socket listener by default pipes messages via stdout. This is an attempt to capture those
```
Comment (yuvalbenshalom) [inline Packs/Slack/Integrations/SlackV3/SlackV3.py:790]: it works?

**Review by yuvalbenshalom — CHANGES_REQUESTED** [review]
see my comments, 
let's have a short talk so I will be able to better understand the changes.

**Review by yuvalbenshalom — APPROVED** [review]

**Pull request merged**